lib/blockchain/chaindb.js

```diff
@@ -888,34 +902,36 @@ class ChainDB {
 
     const start = pruneAfter + 1;
     const end = height - keepBlocks;
-    const b = this.db.batch();
+
+    const blocksBatch = this.blocks.batch();
 
     for (let i = start; i <= end; i++) {
       const hash = await this.getHash(i);
 
       if (!hash)
         throw new Error(`Cannot find hash for ${i}.`);
 
-      b.del(layout.b.encode(hash));
-      b.del(layout.u.encode(hash));
+      blocksBatch.pruneBlock(hash);
+      blocksBatch.pruneUndo(hash);
     }
 
+    // We do blockstore write first, because if something
+    // fails during this batch, then db flag wont be set.
+    // If user just reruns the node prune will restart.
+    await blocksBatch.write();
+
     try {
       options.prune = true;
 
       const flags = ChainFlags.fromOptions(options);
       assert(flags.prune);
 
-      b.put(layout.O.encode(), flags.encode());
-
-      await b.write();
+      await this.db.put(layout.O.encode(), flags.encode());
     } catch (e) {
       options.prune = false;
       throw e;
     }
 
-    await this.db.compactRange();
```

Comment:
I think this was here to compact the db at open, is this not required anymore?

Reply:
Is see that this has moved to migrations. I think a note might make this more clear.

Reply:
This was here for pruning. For some cases, it could prune more than a single block. E.g. with retroactive pruning it could prune many blocks - in that case compactRange was necessary. So migration compactRange is not exactly replacement for this, here we no longer remove data from level db so compactRange loses it's purpose here.

Reply:
OK, thanks that makes sense.